Mathlib/Order/KonigLemma.lean

```diff
@@ -0,0 +1,150 @@
+    [GradeMinOrder ℕ α] [OrderBot α] [Infinite α] (hfin : ∀ (a : α), {x | a ⋖ x}.Finite) :
+    ∃ f : ℕ ↪o α, f 0 = ⊥ ∧ (∀ i, f i ⋖ f (i+1)) ∧ ∀ i, grade ℕ (f i) = i := by
+  obtain ⟨f, h0, hf⟩ := exists_orderEmbedding_covby_of_forall_covby_finite_of_bot hfin
+  refine ⟨f, h0, hf, fun i ↦ ?_⟩
+  induction' i with i ih; simp [h0]
+  simpa [Nat.covBy_iff_succ_eq, ih, eq_comm] using CovBy.grade ℕ <| hf i
+
+end Sequence
+
+section Graded
+
+/-- A formulation of Kőnig's infinity lemma, useful in applications.
+Given a sequence `α 0, α 1, ...` of nonempty types with `α 0` a subsingleton,
+and a well-behaved family of projections `π : α j → α i` for all `i ≤ j`,
+if each term in each `α i` is the projection of only finitely many terms in `α (i+1)`,
+then we can find a sequence `(f 0 : α 0), (f 1 : α 1), ...`
+where `f i` is the projection of `f j` for all `i ≤ j`.
+
+In a typical application, the `α i` are function types with increasingly large domains,
+and `π hij (f : α j)` is the restriction of the domain of `f` to that of `α i`.
+In this case, the sequence given by the lemma is essentially a function whose domain
+is the limit of the `α i`. -/
+theorem exists_seq_forall_proj_of_forall_finite {α : ℕ → Type*} [Subsingleton (α 0)]
+    [∀ i, Nonempty (α i)] (π : {i j : ℕ} → (hij : i ≤ j) → α j → α i)
+    (π_refl : ∀ ⦃i⦄ (a : α i), π rfl.le a = a)
+    (π_trans : ∀ ⦃i j k⦄ (hij : i ≤ j) (hjk : j ≤ k) a, π hij (π hjk a) = π (hij.trans hjk) a)
+    (hfin : ∀ i a, {b : α (i+1) | π (Nat.le_add_right i 1) b = a}.Finite) :
+    ∃ f : (i : ℕ) → α i, ∀ ⦃i j⦄ (hij : i ≤ j), π hij (f j) = f i := by
+
+  set αs := (i : ℕ) × α i
+
+  let _ : PartialOrder αs := {
+    le := fun a b ↦ ∃ h, π h b.2 = a.2
+    le_refl := fun a ↦ ⟨rfl.le, π_refl _⟩
+    le_trans := fun _ _ c h h' ↦ ⟨h.1.trans h'.1, by rw [← π_trans h.1 h'.1 c.2, h'.2, h.2]⟩
+    le_antisymm := by
+      rintro ⟨i, a⟩ ⟨j, b⟩ ⟨hij : i ≤ j, hab : π hij b = a⟩ ⟨hji : j ≤ i, hba : π hji a = b⟩
+      obtain rfl := hij.antisymm hji
+      rw [show a = b by rwa [π_refl] at hba] }
+
+  have hcovby : ∀ {a b : αs}, a ⋖ b ↔ a ≤ b ∧ a.1 + 1 = b.1 := by
+    simp only [covBy_iff_lt_and_eq_or_eq, lt_iff_le_and_ne, ne_eq, Sigma.forall, and_assoc,
+      and_congr_right_iff, or_iff_not_imp_left]
+    rintro i a j b ⟨h : i ≤ j, rfl : π h b = a⟩
+    refine ⟨fun ⟨hne, h'⟩ ↦ ?_, ?_⟩
+    · have hle' : i + 1 ≤ j := h.lt_of_ne <| by rintro rfl; simp [π_refl] at hne
+      exact congr_arg Sigma.fst <| h' (i+1) (π hle' b) ⟨by simp, by rw [π_trans]⟩ ⟨hle', by simp⟩
+        (fun h ↦ by apply_fun Sigma.fst at h; simp at h)
+    rintro rfl
+    refine ⟨fun h ↦ by apply_fun Sigma.fst at h; simp at h, ?_⟩
+    rintro j c ⟨hij : i ≤ j, hcb : π _ c = π _ b⟩ ⟨hji : j ≤ i + 1, rfl : π hji b = c⟩ hne
+    replace hne := show i ≠ j by rintro rfl; exact hne rfl
+    obtain rfl := hji.antisymm (hij.lt_of_ne hne)
+    rw [π_refl]
+
+  have _ : IsStronglyAtomic αs := by
```

Comment:
A small style note you can ignore is that `have` lots you drop the identifier completely:
```suggestion
  have : IsStronglyAtomic αs := by
```

Reply:
I use the underscore because I don't like the brightly coloured `this` in the infoview for instances. Is the underscore considered bad style?